syncer/error.go

```diff
@@ -70,3 +75,53 @@ func originError(err error) error {
 	}
 	return err
 }
+
+// handleSpecialDDLError handles special errors for DDL execution.
+// it only ignore `invalid connection` error (timeout or other causes) for `ADD INDEX` now.
+// `invalid connection` means some data already sent to the server,
+// and we assume that the whole SQL statement has already sent to the server for this error.
+// if we have other methods to judge the DDL dispatched but timeout for executing, we can update this method.
+// NOTE: we must ensure other PK/UK exists for correctness.
+// NOTE: when we are refactoring the shard DDL algorithm, we also need to consider supporting non-blocking `ADD INDEX`.
+func (s *Syncer) handleSpecialDDLError(tctx *tcontext.Context, err error, ddls []string, index int, conn *DBConn) error {
+	// must ensure only the last statement executed failed with the `invalid connection` error
+	if len(ddls) == 0 || index != len(ddls)-1 || errors.Cause(err) != mysql.ErrInvalidConn {
+		return err // return the original error
+	}
+
+	parser2, err2 := s.fromDB.getParser(s.cfg.EnableANSIQuotes)
+	if err2 != nil {
+		return err // return the original error
+	}
+
+	ddl2 := ddls[index]
+	stmt, err2 := parser2.ParseOneStmt(ddl2, "", "")
+	if err2 != nil {
+		return err // return the original error
+	}
+
+	handle := func() {
+		tctx.L().Warn("ignore special error for DDL", zap.String("DDL", ddl2), log.ShortError(err))
+		err2 := conn.resetConn(tctx) // also reset the `invalid connection` for later use.
+		if err2 != nil {
+			tctx.L().Warn("reset connection failed", log.ShortError(err2))
+		}
+	}
+
+	switch v := stmt.(type) {
+	case *ast.AlterTableStmt:
+		// ddls should be split with only one spec
+		if len(v.Specs) > 1 {
+			return err
+		} else if v.Specs[0].Tp == ast.AlterTableAddConstraint {
+			handle()
+			return nil // ignore the error
```

Comment:
why ignore error when tp is AlterTableAddConstraint?

Reply:
`ast.AlterTableAddConstraint` represents `ALTER TABLE xxx ADD INDEX yyy`, and `ast.CreateIndexStmt` represents `CREATE INDEX yyy ON xxx`.

Reply:
it may be `add column` or other type?

Reply:
It only can be index and foreign key now in TiDB (https://github.com/pingcap/tidb/blob/3cdea0dfdf28197ee65545debce8c99e6d2945e3/ddl/ddl_api.go#L1929-L1948).

do you think also need to check `spec.Constraint.Tp` in DM?

Reply:
ok, LGTM

Reply:
add more check in https://github.com/pingcap/dm/pull/336/commits/e4b29b44890b82a951852e0ad33ee14e727087b7.